### PR TITLE
set raftstore thread priority

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Two requirement:
1) user needs `CAP_SYS_NICE` capability.
2) only support linux platform.
or, nothing will happen.

In order to add `CAP_SYS_NICE` for user tidb, please add following two lines to /etc/security/limits.conf:
tidb        soft        nice          -1
tidb        hard        nice          -1

### CODE REVIEW

File: src/util/sys/mod.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#[cfg(target_os = "linux")]
+pub fn adjust_priority(t: AdjustPriType) {
+    unsafe {
+        let pid = libc::pthread_self();
+        let pri = match t {
+            AdjustPriType::Incr => -1,
+            AdjustPriType::Normal => 0,
+            AdjustPriType::Desc => 1,
+        };
+        libc::setpriority(libc::PRIO_PROCESS as u32, pid as u32, pri);
```
Comment (ngaut) [inline src/util/sys/mod.rs]: Any way to check if the setting successful or not?

File: src/util/sys/mod.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+use std::io::Error;
+
+pub enum AdjustPriType {
+    Incr,
+    Normal,
+    Desc,
+}
+
+#[cfg(target_os = "linux")]
+pub fn adjust_priority(t: AdjustPriType) {
```
Comment (siddontang) [inline src/util/sys/mod.rs]: seem you can provide two `incr_priority` and `decr_priority` functions

**Comment by siddontang** [comment]
please add test

File: src/util/sys/mod.rs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    unsafe {
+        let pid = libc::pthread_self();
+        let pri = match t {
+            AdjustPriType::Incr => -1,
+            AdjustPriType::Normal => 0,
+            AdjustPriType::Desc => 1,
+        };
+        if libc::setpriority(libc::PRIO_PROCESS as u32, pid as u32, pri) != 0 {
```
Comment (siddontang) [inline src/util/sys/mod.rs]: the setpriority can support the nice value like in range [-20, 19], see https://linux.die.net/man/2/setpriority, why do we only have -1, 0 and 1 here? 

How can we set different nice values for other threads?

Comment (zhangjinpeng1987) [reply src/util/sys/mod.rs]: Do we have such demand? I just don't want to make it too complicated.

File: src/util/sys/mod.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#[cfg(target_os = "linux")]
+use libc;
+#[cfg(target_os = "linux")]
+use std::io::Error;
+
+pub enum AdjustPriType {
```
Comment (siddontang) [inline src/util/sys/mod.rs]:
```
pub enum PriorityType {
    High,
    Normal,
    Low
}

pub fn set_priority()
```
IMO, adjust with Incr and Decr confuses me, calling `adjust` witch Incr doesn't mean increasing priority each time, so using `set` with High or Low here is better here.

File: src/util/sys/mod.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        #[test]
+        #[cfg(target_os = "linux")]
+        fn test_set_priority() {
+            // priority is a value in range -20 to 19, the default priority
+            // is 0, lower priorities cause more favorable scheduling.
+            assert_eq!(get_priority().unwrap(), 0);
+            set_priority(10).unwrap();
+            assert_eq!(get_priority().unwrap(), 10);
+
+            // only users which have `SYS_NICE_CAP` capability can priority.
```
Comment (siddontang) [inline src/util/sys/mod.rs]: can increase priority

File: src/util/sys/mod.rs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            let ret = set_priority(-10);
```
Comment (siddontang) [inline src/util/sys/mod.rs]: use HIGH_PRI here

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/util/sys/mod.rs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    pub fn get_priority() -> Result<i32, Error> {
+        unsafe {
+            let pid = libc::getpid();
+            let ret = libc::getpriority(libc::PRIO_PROCESS as u32, pid as u32);
+            if ret == -1 {
+                let e = Error::last_os_error();
+                if e.kind() != ErrorKind::Other {
```
Comment (BusyJay) [inline src/util/sys/mod.rs]: Why ignore `Other`?

Comment (zhangjinpeng1987) [reply src/util/sys/mod.rs]: https://github.com/arangodb/arangodb/blob/devel/arangod/RocksDBEngine/RocksDBThrottle.cpp

File: src/util/sys/mod.rs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+pub const HIGH_PRI: i32 = -1;
+
+#[cfg(target_os = "linux")]
+pub mod pri {
+    use libc;
+    use std::io::Error;
+
+    pub fn set_priority(pri: i32) -> Result<(), Error> {
+        unsafe {
+            let pid = libc::syscall(libc::SYS_gettid);
```
Comment (BusyJay) [inline src/util/sys/mod.rs]: Should name it `tid`.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/util/sys/mod.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        #[test]
+        fn test_set_priority() {
+            // priority is a value in range -20 to 19, the default priority
+            // is 0, lower priorities cause more favorable scheduling.
+            assert_eq!(get_priority().unwrap(), 0);
+            set_priority(10).unwrap();
+            assert_eq!(get_priority().unwrap(), 10);
+
+            // only users which have `SYS_NICE_CAP` capability can increase priority.
```
Comment (BusyJay) [inline src/util/sys/mod.rs]: s/which/who/

**Comment by siddontang** [comment]
how can we ensure the higher priority can reducing the switching frequency of raftstore?

**Pull request merged**